Added default constructor

M66B opened the pull request.

M66B committed:
Added default constructor

M66B commented:
With a default constructor, the class can be used as an esphome global variable, and the stream being set with *on_boot*.

```yml
globals:
  - id: mtp
    type: MTP40F

on_boot:
  id(mtp).setStream(new UARTStream(new UARTDevice(id(uart_bus))));
```

Since esphome has no streams (yet?), I'm using this to create a stream for the MTP40F class:

```cpp
#include "esphome.h"

using namespace esphome;

class UARTStream : public Stream {
protected:
  UARTDevice *dev;

 public:
  UARTStream(UARTDevice *device) {
    this->dev = device;
  }

  size_t write(uint8_t x) {
    return this->dev->write(x);
  }

  int available() {
    return this->dev->available();
  }

  int read() {
    return this->dev->read();
  }

  int peek() {
    return this->dev->peek();
  }
};
```

RobTillaart commented:
@M66B 

Thanks for the PR, will merge as it all looks straightforward.

Some questions
1. What is the rationale, why do you need the extra constructor?

2. Should setStream() call init() too?

3. If no stream is set should type be e.g. N,  to indicate the stream has not been set yet?

RobTillaart commented:
Crosspost,
That answers Q1. 
Thanks,

M66B commented:
> 1. What is the rationale, why do you need the extra constructor?

Please see above.

> 2. Should setStream() call init() too?

It would be better indeed, but for my use case it doesn't matter.

> 3. If no stream is set should type be e.g. N,  to indicate the stream has not been set yet?

Good point as well.

Let me see what I can do.

M66B committed:
Check if stream is set

M66B commented:
A new commit is addressing points 2 and 3.

Note that somebody could have 'cheated' with the original solution too by passing a null pointer to the constructor ;-)

RobTillaart commented:
> Note that somebody could have 'cheated' with the original solution too by passing a null pointer to the constructor ;-)

Definitely very true, or just any pointer casted to a Stream pointer.
The new constructor just increased the chances, your solution catches the NULL pointer so well solved.

Do you expect more changes, if not I will merge later today.

M66B commented:
I don't expect further changes. Basically, it is a pretty trivial, but still useful (IMHO), addition.

It would be nice if the version would be updated as well and if the library was being published as Arduino library. The version update is something I can take care of if you wish.

M66B committed:
Bump to 0.2.1

RobTillaart commented:
Thanks!
There are two small edits open "needed" for a new release
- keywords.txt
- two lines in the readme.md file in the interface section.

M66B committed:
Updated documentation

M66B commented:
Done, with in addition documentation of the new error code.

RobTillaart commented:
OK, all looks good 
so I will merge and release
Thanks again for the PR and the additions!

RobTillaart commented:
@M66B 

0.2.1 released, expect it will be updated in the Arduino libs within a day (often faster).
I will also update the latest in platformIO as that is not automatic

M66B commented:
Thanks, Rob.

For the record: this sensor has been working reliably for me since July 2023.

RobTillaart commented:
PlatformIO has been updated, 
Created an issue to check to sync the MTP40C lib as these are family.

M66B commented:
Arduino libraries have been updated as well now.
